File: 백준/Gold/13023. ABCDE/ABCDE.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

public class Main {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringTokenizer st;
	static int n,m,a,b,ok;
	static List<List<Integer>> g;
	static int[] visit;
	public static void main(String[] args) throws Exception{
		st = new StringTokenizer(br.readLine());
		n = Integer.parseInt(st.nextToken());
		g = new ArrayList<>();
		for (int i=0; i<n; i++) {
			g.add(new ArrayList<>());
		}
		visit = new int[n];
		m = Integer.parseInt(st.nextToken());
		for (int i=0; i<m; i++) {
			st = new StringTokenizer(br.readLine());
			a = Integer.parseInt(st.nextToken());
			b = Integer.parseInt(st.nextToken());
			g.get(a).add(b);
			g.get(b).add(a);
		}
		ok = 0;
		for (int j=0; j<n; j++) {
			dfs(j,0);
			if (ok==1)
				break;
		}
		System.out.println(ok);
	}
	public static void dfs(int node, int cnt) {
		if (cnt==4 || ok==1) {
			ok=1;
			return;
		}
		visit[node] = 1;
		for (int next: g.get(node)) {
			if (visit[next]==0) {
				dfs(next, cnt+1);
			}
		}
		visit[node] = 0;
	}
}

File: 백준/Gold/13023. ABCDE/ABCDE.py
import sys
input = sys.stdin.readline

def dfs(now, cnt):
    global result
    if cnt >=4:
        result = 1
        return
    for next in g[now]:
        if visit[next] == 0:
            visit[next] = 1
            dfs(next,cnt+1)
            visit[next] = 0

n,m = map(int, input().split())
visit = [0]*n
g = [[] for _ in range(n)]
for i in range(m):
    a,b = map(int, input().split())
    g[a].append(b)
    g[b].append(a)

result = 0
for j in range(n):
    visit[j] = 1
    dfs(j,0)
    if result == 1:
        break
    visit[j] = 0
print(result)